**Modules**

In [6]:
import pandas as pd
import numpy as np
import json
import gzip
import gc

### Selecting Books With `ratings_count > 15`

**Function to parse the specific fields of the json data**

In [7]:
def parse_fields(line):
    data = json.loads(line)
    return {
        'isbn' : data['isbn'],
        'text_reviews_count' : data['text_reviews_count'],
        'series' : data['series'],
        'country_code' : data['country_code'],
        'language_code' : data['language_code'],
        'asin' : data['asin'],
        'is_ebook' : data['is_ebook'],
        'average_rating' : data['average_rating'],
        'kindle_asin' : data['kindle_asin'],
        'similar_books' : data['similar_books'],
        'description' : data['description'],
        'format' : data['format'],
        'link' : data['link'],
        'authors' : data['authors'],
        'publisher' : data['publisher'],
        'num_pages' : data['num_pages'],
        'publication_day' : data['publication_day'],
        'isbn13' : data['isbn13'],
        'publication_month' : data['publication_month'],
        'publication_year' : data['publication_year'],
        'url' : data['url'],
        'image_url' : data['image_url'],
        'book_id' : data['book_id'],
        'ratings_count' : data['ratings_count'],
        'title_without_series' : data['title_without_series']
    }

**We will go through all the books from the dataset, line by line till the end using an infinite loop**
- We will parse the details of only those books which has got `ratings_count >= 15`

In [8]:
books = []

with gzip.open("../Initial/books.json.gz") as f:
    while True:
        # reading the line
        line = f.readline()

        # we will break the infinite loop when we reach the end of the dataset file
        if not line:
            break
        
        # parsing the line
        fields = parse_fields(line)
        
        # trying to convert ratings_count into integer
        try:
            ratings_count = int(fields["ratings_count"])
        except ValueError:
            continue

        # we will consider only those books which has more than 15 ratings
        if ratings_count > 15:
            books.append(fields)

**Total number of books**

In [9]:
len(books)

1308957

**Insight**
- There are 1,308,957 books after filtering through the criteria `ratings_count > 15`

In [10]:
books[0]

{'isbn': '',
 'text_reviews_count': '7',
 'series': ['189911'],
 'country_code': 'US',
 'language_code': 'eng',
 'asin': 'B00071IKUY',
 'is_ebook': 'false',
 'average_rating': '4.03',
 'kindle_asin': '',
 'similar_books': ['19997',
  '828466',
  '1569323',
  '425389',
  '1176674',
  '262740',
  '3743837',
  '880461',
  '2292726',
  '1883810',
  '1808197',
  '625150',
  '1988046',
  '390170',
  '2620131',
  '383106',
  '1597281'],
 'description': 'Omnibus book club edition containing the Ladies of Madrigyn and the Witches of Wenshar.',
 'format': 'Hardcover',
 'link': 'https://www.goodreads.com/book/show/7327624-the-unschooled-wizard',
 'authors': [{'author_id': '10333', 'role': ''}],
 'publisher': 'Nelson Doubleday, Inc.',
 'num_pages': '600',
 'publication_day': '',
 'isbn13': '',
 'publication_month': '',
 'publication_year': '1987',
 'url': 'https://www.goodreads.com/book/show/7327624-the-unschooled-wizard',
 'image_url': 'https://images.gr-assets.com/books/1304100136m/7327624.jpg',

**Creating a DataFrame from the list of dictionaries**

In [11]:
items = pd.DataFrame.from_dict(books)

**Rows and Columns of the DataFrame**

In [12]:
print(f"Rows: {items.shape[0]}")
print(f"Columns: {items.shape[1]}")

Rows: 1308957
Columns: 25


**We don't need the `books` object anymore**
- Deleting the object from namespace and forcing Python for garbage collection (if required)
- The `del` method only removes the variable from the namespace and it does not remove the variable from the memory space
- To clear the variable from memory we can use the `gc.collect()` method

In [13]:
del(books)

In [14]:
gc.collect()

0

### Creating Modified Title to Minimize Search Space

**Removing characters apart from A-z, a-z and 0-9 and Space from `title_without_series` and saving into new `mod_title` column**

In [15]:
items["mod_title"] = items["title_without_series"].str.replace("[^a-zA-Z0-9 ]", "", regex=True)

**Make `mod_title` lower case**

In [16]:
items["mod_title"] = items["mod_title"].str.lower()

**Replacing any extra spaces with a single space**

In [17]:
items["mod_title"] = items["mod_title"].str.replace("\s+", " ", regex=True)

**Keeping only those records where the `mod_title` is not empty or has a length > 0**

In [18]:
items = items[items["mod_title"].str.len() > 0]

In [19]:
len(items) 

1302659

**Insight**
- We can observe a sligt reduction in count which is expected
  -  `(1308957 -> 1302659)`

#### `mod_title` Column Analysis - Important For Search Engine

**Calculate ratio of `len(title_without_series) and len(mod_title)`**
- This will give us a measure of truncation

In [20]:
items["actualT_modT_ratio"] = items["title_without_series"].str.len() / items["mod_title"].str.len()
items.head()

isbn text_reviews_count    series country_code language_code  \
0                              7  [189911]           US           eng   
1  0743294297               3282        []           US           eng   
2  1599150603                  7        []           US                 
3  1934876569                  6  [151854]           US                 
4  0922915113                 39        []           US                 

         asin is_ebook average_rating kindle_asin  \
0  B00071IKUY    false           4.03               
1                false           3.49  B002ENBLOK   
2                false           4.13  B00DU10PUG   
3                false           4.22               
4                false           3.81  B00AFYVB8Q   

                                       similar_books  ...         isbn13  \
0  [19997, 828466, 1569323, 425389, 1176674, 2627...  ...                  
1  [6604176, 6054190, 2285777, 82641, 7569453, 70...  ...  9780743294294   
2                                                 []  ...  9781599150604   
3  [948696, 439885, 274955, 12978730, 372986, 216...  ...  9781934876565   
4  [287151, 1104760, 1172822, 440292, 287082, 630...  ...  9780922915118   

  publication_month publication_year  \
0                               1987   
1                 7             2009   
2                 9             2006   
3                 3             2009   
4                 4             2000   

                                                 url  \
0  https://www.goodreads.com/book/show/7327624-th...   
1  https://www.goodreads.com/book/show/6066819-be...   
2  https://www.goodreads.com/book/show/287141.The...   
3  https://www.goodreads.com/book/show/6066812-al...   
4  https://www.goodreads.com/book/show/287149.The...   

                                           image_url  book_id ratings_count  \
0  https://images.gr-assets.com/books/1304100136m...  7327624           140   
1  https://s.gr-assets.com/assets/nophoto/book/11...  6066819         51184   
2  https://s.gr-assets.com/assets/nophoto/book/11...   287141            46   
3  https://images.gr-assets.com/books/1316637798m...  6066812            98   
4  https://images.gr-assets.com/books/1328768789m...   287149           986   

                                title_without_series  \
0  The Unschooled Wizard (Sun Wolf and Starhawk, ...   
1                               Best Friends Forever   
2                      The Aeneid for Boys and Girls   
3  All's Fairy in Love and War (Avalon: Web of Ma...   
4                               The Devil's Notebook   

                                          mod_title actualT_modT_ratio  
0    the unschooled wizard sun wolf and starhawk 12           1.108696  
1                              best friends forever           1.000000  
2                     the aeneid for boys and girls           1.000000  
3  alls fairy in love and war avalon web of magic 8           1.125000  
4                               the devils notebook           1.052632  

[5 rows x 27 columns]

**Looking at `actualT_modT_ratio` values for actual titles having very short length `< 5`**

In [21]:
items.loc[items["title_without_series"].str.len() < 5]

isbn text_reviews_count    series country_code language_code  \
21       1597371289                  8        []           US           eng   
226                                  7  [297710]           US           eng   
346                                  5        []           US                 
349                                  2        []           US                 
436                                  5        []           US           eng   
...             ...                ...       ...          ...           ...   
1307639                             16        []           US                 
1307948  0375759603                 28        []           US                 
1307959  0547728247                  3        []           US           eng   
1308509  1626391912                 18        []           US           eng   
1308782  0413519503                  4        []           US                 

               asin is_ebook average_rating kindle_asin  \
21                     false           3.99  B0083Z3O8Y   
226                    false           3.81               
346                    false           3.33               
349      B001EL6RSS     true           3.38  B001EL6RSS   
436                    false           3.55               
...             ...      ...            ...         ...   
1307639  B01F293HAS     true           4.07               
1307948                false           3.29  B000FC1I6S   
1307959                 true           4.10               
1308509                false           3.75  B00ND8SUT4   
1308782                false           3.51  B00G99N4AO   

                                             similar_books  ...  \
21       [31242, 374380, 20564, 383206, 7891, 6335178, ...  ...   
226      [356205, 266753, 379260, 155159, 45684, 888887...  ...   
346                                                     []  ...   
349      [1618765, 12934562, 130, 61125, 38524, 7714891...  ...   
436                                                     []  ...   
...                                                    ...  ...   
1307639  [26312997, 25897916, 25897871, 27311743, 24611...  ...   
1307948  [3433301, 118192, 124893, 1001309, 1136604, 22...  ...   
1307959  [219107, 76740, 9682235, 18952, 60146, 845501,...  ...   
1308509                                                 []  ...   
1308782  [1975642, 1047906, 589143, 18226469, 1103629, ...  ...   

                isbn13 publication_month publication_year  \
21       9781597371285                 9             2005   
226                                                         
346      9789949301317                               2011   
349                                                         
436                                                         
...                ...               ...              ...   
1307639                                                     
1307948  9780375759604                               2002   
1307959  9780547728247                 4             2012   
1308509  9781626391918                 9             2014   
1308782  9780413519504                 4             1983   

                                                       url  \
21        https://www.goodreads.com/book/show/3209316-emma   
226      https://www.goodreads.com/book/show/17794787-legs   
346       https://www.goodreads.com/book/show/13325505-7x7   
349      https://www.goodreads.com/book/show/19222476-send   
436      https://www.goodreads.com/book/show/18042275-skin   
...                                                    ...   
1307639  https://www.goodreads.com/book/show/30125838-grit   
1307948    https://www.goodreads.com/book/show/904266.Fury   
1307959  https://www.goodreads.com/book/show/16555672-ubik   
1308509  https://www.goodreads.com/book/show/20702686-jolt   
1308782    https://www.goodreads.com/book/show/315118.Lear   

                                                 image_url   book_id  \
21

**Looking at the statistics of `actualT_modT_ratio` column while `title_without_series < 5`**

In [22]:
items.loc[items["title_without_series"].str.len() < 5, "actualT_modT_ratio"].describe()

count    8253.000000
mean        1.032736
std         0.165538
min         1.000000
25%         1.000000
50%         1.000000
75%         1.000000
max         4.000000
Name: actualT_modT_ratio, dtype: float64

In [23]:
np.percentile(items.loc[items["title_without_series"].str.len() < 5, "actualT_modT_ratio"].values,94)

1.0

In [24]:
np.percentile(items.loc[items["title_without_series"].str.len() < 5, "actualT_modT_ratio"].values,97)

1.3333333333333333

In [25]:
np.percentile(items.loc[items["title_without_series"].str.len() < 5, "actualT_modT_ratio"].values,98)

1.5

In [26]:
np.percentile(items.loc[items["title_without_series"].str.len() < 5, "actualT_modT_ratio"].values,99)

2.0

**Insight**
- Most of the distribution with the given criteria `(almost 99%)` is having `actualT_modT_ratio <= 2`
- So, we can consider `2` as our threshold while `title_without_series < 5`

**Based on the above parameters looking at data when `len(title_without_series) < 5` and `actualT_modT_ratio > 2`**

In [27]:
temp_df = items.loc[items["title_without_series"].str.len() < 5]
temp_df.loc[temp_df["actualT_modT_ratio"] > 2].sample(10)

isbn text_reviews_count                            series  \
918636                              11                                []   
270075   9758518089                  2  [179007, 179001, 179011, 889174]   
103179   9953215073                 12                                []   
1095472                             14                          [243751]   
113279                              37                                []   
983859   275602595X                 11                                []   
658456   0801846684                 24                                []   
92498                                2                                []   
380971   9866562131                  1                                []   
75566                                8                                []   

        country_code language_code asin is_ebook average_rating kindle_asin  \
918636            US           pes         false           4.23               
270075            US           tur         false           4.23               
103179            US           ara         false           3.64               
1095472           US           ara         false           3.59               
113279            US                       false           3.91               
983859            US           fre         false           4.01               
658456            US           eng         false           4.27               
92498             US           fin         false           3.73               
380971            US           zho         false           3.96               
75566             US           ara         false           3.62               

                                             similar_books  ...  \
918636                                                  []  ...   
270075   [291693, 978999, 362055, 445761, 362053, 52958...  ...   
103179                                                  []  ...   
1095472                        [6037124, 9553204, 6445626]  ...   
113279   [12907023, 2334969, 10872535, 6652169, 522888,...  ...   
983859                                                  []  ...   
658456   [385120, 409241, 150347, 385672, 405100, 33094...  ...   
92498                                                   []  ...   
380971                                                  []  ...   
75566                                                   []  ...   

                isbn13 publication_month publication_year  \
918636   9786009545001                               2015   
270075   9789758518081                12             2001   
103179                                 6             2011   
1095472                                7             2010   
113279                                 3             2014   
983859   9782756025957                 9             2011   
658456   9780801846687                10             1993   
92498    9789529887545                               2008   
380971   9789866562136                 1             2009   
75566                                 12             2014   

                                                    url  \
918636     https://www.goodreads.com/book/show/28930830   
270075   https://www.goodreads.com/book/show/19478121-g   
103179     https://www.goodreads.com/book/show/13610925   
1095472     https://www.goodreads.com/book/show/6279485   
113279   https://www.goodreads.com/book/show/21460333-o   
983859   https://www.goodreads.com/book/show/12873782-3   
658456   https://www.goodreads.com/book/show/375182._A_   
92498     https://www.goodreads.com/book/show/6272309-j   
380971      https://www.goodreads.com/book/show/9967288   
75566    https://www.goodreads.com/book/show/24012899-3   

                                                 image_url   book_id  \
918636   https://images.gr-assets.com/books/1454571341m...  28930830   
270075   https://images.gr-assets.com/books/1387111296m...  19478121   
103179   https://images.gr-asset

**We can get the statistics of `mod_title` length when (`actualT_modT_ratio > 2`)**

In [28]:
items.loc[(items["actualT_modT_ratio"] > 2), "mod_title"].str.len().describe()

count    42709.000000
mean         1.512655
std          2.431791
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max         45.000000
Name: mod_title, dtype: float64

In [29]:
for i in range(85,100):
    print(i, "=>" ,np.percentile(items.loc[(items["actualT_modT_ratio"] > 2), "mod_title"].str.len(),i))

85 => 1.0
86 => 1.0
87 => 1.0
88 => 1.0
89 => 2.0
90 => 2.0
91 => 2.0
92 => 2.0
93 => 2.0
94 => 2.0
95 => 3.0
96 => 4.0
97 => 6.0
98 => 10.0
99 => 15.0


**Insight**
- We cannot use only `actualT_modT_ratio > 2` as our criteria because we might loose some data with `mod_title > 4`

**We will add `mod_title` length threshold as well and will also adjust the value of `actualT_modT_ratio`**
- Based on observation testing out some values

In [48]:
items.loc[(items["actualT_modT_ratio"] > 2.3) & (items["mod_title"].str.len() > 6)]

isbn text_reviews_count                            series  \
727                                 33                          [691556]   
4253                                 5                                []   
4620                                 1                          [769261]   
5034                                 1                                []   
6048                                10                          [273034]   
...             ...                ...                               ...   
1301558  9542813007                  4                                []   
1305982                              3  [787876, 773429, 483372, 755577]   
1307757                              2                                []   
1307790                              2                          [772452]   
1308893  0966696192                  1                                []   

        country_code language_code asin is_ebook average_rating kindle_asin  \
727               US           gre         false           3.88               
4253              US           rus         false           3.33               
4620              US           ben         false           3.41               
5034              US           ara         false           4.08               
6048              US           gre         false           3.87               
...              ...           ...  ...      ...            ...         ...   
1301558           US            --         false           3.41               
1305982           US           kat         false           4.12               
1307757           US           tha         false           3.73               
1307790           US           ben         false           4.42               
1308893           US           rus         false           4.26               

                                             similar_books  ...  \
727      [18102876, 18052948, 21350894, 18135589, 24909...  ...   
4253                                                    []  ...   
4620                                                    []  ...   
5034                                                    []  ...   
6048     [15704485, 13596809, 13612739, 15739921, 16070...  ...   
...                                                    ...  ...   
1301558                                                 []  ...   
1305982  [7552385, 12973980, 20924715, 131332, 16028305...  ...   
1307757  [5202, 281285, 639374, 953345, 1319826, 287505...  ...   
1307790                                                 []  ...   
1308893                                                 []  ...   

                isbn13 publication_month publication_year  \
727      9789603648482                 4             2015   
4253     9785916571431                               2011   
4620                                                        
5034                                                 2002   
6048     9789601646046                10             2012   
...                ...               ...              ...   
1301558  9789542813002                 5             2013   
1305982  9789941467288                               2016   
1307757                                                     
1307790                                                     
1308893                                              2001   

                                                       url  \
727           https://www.goodreads.com/book/show/25204099   
4253     https://www.goodreads.com/book/show/10143804--...   
4620          https://www.goodreads.com/book/show/25454679   
5034     https://www.goodreads.com/book/show/9678144-19...   
6048          https://www.goodreads.com/book/show/16049765   
...                                                    ...   
1301558  https://www.goodreads.com/book/show/17924291-t...   
1305982  https://www.goodreads.com/book/show/31699289---ii   
1307757  https://www.goodreads.com/book/show/9570864-th..

In [49]:
items.loc[(items["actualT_modT_ratio"] > 2.3) & (items["mod_title"].str.len() < 7), "mod_title"].str.len().describe()

count    41523.000000
mean         1.150591
std          0.576530
min          1.000000
25%          1.000000
50%          1.000000
75%          1.000000
max          6.000000
Name: mod_title, dtype: float64

In [50]:
for i in range(85,100):
    print(i, "=>" ,np.percentile(items.loc[(items["actualT_modT_ratio"] > 2.3) & (items["mod_title"].str.len() < 7), "mod_title"].str.len(),i))

85 => 1.0
86 => 1.0
87 => 1.0
88 => 1.0
89 => 1.0
90 => 1.0
91 => 2.0
92 => 2.0
93 => 2.0
94 => 2.0
95 => 2.0
96 => 2.0
97 => 3.0
98 => 3.0
99 => 4.0


In [55]:
for i in range(0,10):
    print(i, "=>" ,np.percentile(items.loc[(items["actualT_modT_ratio"] > 2.4) & (items["mod_title"].str.len() < 7), "actualT_modT_ratio"],i))

0 => 2.5
1 => 4.5
2 => 6.2
3 => 7.0
4 => 8.0
5 => 8.0
6 => 9.0
7 => 9.0
8 => 9.0
9 => 9.0


**Insight**
- We have fixed the threshold
  - `actualT_modT_ratio > 2.3` and `mod_title < 7`

In [ ]:
# execution stop point

10/0

**Index of the rows to delete**

In [56]:
index_to_del = items.loc[(items["actualT_modT_ratio"] > 2.3) & (items["mod_title"].str.len() < 7)].index
index_to_del

Index([     28,      37,      38,      43,      65,      72,      85,     108,
           118,     175,
       ...
       1308680, 1308686, 1308691, 1308697, 1308789, 1308850, 1308854, 1308881,
       1308931, 1308940],
      dtype='int64', length=41523)

**Dropping the rows**

In [57]:
items.drop(index=index_to_del, inplace=True)

**Dropping the `actualT_modT_ratio` column**

In [58]:
items.drop(columns="actualT_modT_ratio", inplace=True)

In [59]:
len(items)

1261136

In [ ]:
# execution stop point

# 10/0

**Insight**
- We can observe a sligt reduction in count which is expected
  -  `(1,308,957 -> 1,302,659 -> 1,261,136)`

**Exporting the Data as Compressed JSON After Reducing Search Space**

In [60]:
with gzip.open('../Processed/books_p0.json.gz', 'wt', encoding='utf-8') as file:
    file.write(items.to_json(orient='records', lines=True))

In [ ]:
# execution stop point

# 10/0

### Testing The Exported File

In [ ]:
# testing the file  => opening gzip file streaming fashion

with gzip.open('../Processed/books_p0.json.gz') as file:
    line = file.readline()

line

b'{"isbn":"","text_reviews_count":"7","series":["189911"],"country_code":"US","language_code":"eng","asin":"B00071IKUY","is_ebook":"false","average_rating":"4.03","kindle_asin":"","similar_books":["19997","828466","1569323","425389","1176674","262740","3743837","880461","2292726","1883810","1808197","625150","1988046","390170","2620131","383106","1597281"],"description":"Omnibus book club edition containing the Ladies of Madrigyn and the Witches of Wenshar.","format":"Hardcover","link":"https:\\/\\/www.goodreads.com\\/book\\/show\\/7327624-the-unschooled-wizard","authors":[{"author_id":"10333","role":""}],"publisher":"Nelson Doubleday, Inc.","num_pages":"600","publication_day":"","isbn13":"","publication_month":"","publication_year":"1987","url":"https:\\/\\/www.goodreads.com\\/book\\/show\\/7327624-the-unschooled-wizard","image_url":"https:\\/\\/images.gr-assets.com\\/books\\/1304100136m\\/7327624.jpg","book_id":"7327624","ratings_count":"140","title_without_series":"The Unschooled Wi

In [ ]:
json.loads(line)

{'isbn': '',
 'text_reviews_count': '7',
 'series': ['189911'],
 'country_code': 'US',
 'language_code': 'eng',
 'asin': 'B00071IKUY',
 'is_ebook': 'false',
 'average_rating': '4.03',
 'kindle_asin': '',
 'similar_books': ['19997',
  '828466',
  '1569323',
  '425389',
  '1176674',
  '262740',
  '3743837',
  '880461',
  '2292726',
  '1883810',
  '1808197',
  '625150',
  '1988046',
  '390170',
  '2620131',
  '383106',
  '1597281'],
 'description': 'Omnibus book club edition containing the Ladies of Madrigyn and the Witches of Wenshar.',
 'format': 'Hardcover',
 'link': 'https://www.goodreads.com/book/show/7327624-the-unschooled-wizard',
 'authors': [{'author_id': '10333', 'role': ''}],
 'publisher': 'Nelson Doubleday, Inc.',
 'num_pages': '600',
 'publication_day': '',
 'isbn13': '',
 'publication_month': '',
 'publication_year': '1987',
 'url': 'https://www.goodreads.com/book/show/7327624-the-unschooled-wizard',
 'image_url': 'https://images.gr-assets.com/books/1304100136m/7327624.jpg',